src/devices/Board/I2cBusExtensions.cs

```diff
@@ -44,5 +44,112 @@ public static List<int> PerformBusScan(this I2cBus bus, int lowest = 0x3, int hi
 
             return ret;
         }
+
+        /// <summary>
+        /// Performs a scan on the I2C bus, returning the addresses for all connected devices.
+        /// </summary>
+        /// <param name="bus">The bus to scan</param>
+        /// <param name="progress">Progress feedback provider. Receives scan progress in percent</param>
+        /// <param name="lowestAddress">The lowest address to scan. Default 0x03</param>
+        /// <param name="highestAddress">The highest address to scan. Default 0x77</param>
+        /// <returns>A list of bus addresses that are in use, an empty list if no device was found</returns>
+        /// <remarks>This method should never throw an exception. Bus scanning may interfere with normal device operation,
+        /// so this should not be done while devices are being used.</remarks>
+        public static (List<int> FoundDevices, int LowestAddress, int HighestAddress) PerformBusScan(this I2cBus bus, IProgress<float>? progress, int lowestAddress = 0x3, int highestAddress = 0x77)
+        {
+            List<int> addresses = new();
+            int numberOfDevicesToScan = highestAddress - lowestAddress;
+            float currentPercentage = 0;
+            float stepPerDevice = 100.0f / numberOfDevicesToScan;
+            if (lowestAddress < 0)
```

Comment:
NIT: You should probably move all of this argument validation to the top of the method body since you don't want to do any initialization in a failure case.